Ajc/updates to docs for migration (#818)
Qiskit-Extensions/circuit-knitting-toolbox: docs/explanation/index.rst

Overview of circuit cutting
---------------------------

> Circuit cutting is a technique to increase the size of circuits we can run on quantum hardware at the cost of an additional sampling overhead. A larger quantum circuit can be decomposed by cutting its gates or wires, resulting in smaller circuits that can be executed within the constraints of available quantum hardware. The results of these smaller circuits are combined to reconstruct the outcome of the original problem. Circuit cutting can also be used to engineer gates between distant qubits which would otherwise require a large swap overhead.

Reviewer: Being able to cut wires and gates is the same circuit is a key feature of this package.  Here is one way to express this explicitly, but I am open to others.

```suggestion
Circuit cutting is a technique to increase the size of circuits we can run on quantum hardware at the cost of an additional sampling overhead. A larger quantum circuit can be decomposed by cutting its gates, wires, or both, resulting in smaller circuits that can be executed within the constraints of available quantum hardware. The results of these smaller circuits are combined to reconstruct the outcome of the original problem. Circuit cutting can also be used to engineer gates between distant qubits which would otherwise require a large swap overhead.
```

There are two types of cuts: gate cuts and wire cuts.  Gate cuts, also known as "space-like" cuts, exist when the cut goes through a gate operating on two (or more) qubits.  Wire cuts, also known as "time-like" cuts, are direct cuts through a qubit wire; essentially, they are single-qubit identity gates that have been cut into two pieces.  In this package, a wire cut is represented by introducing a new qubit into the circuit and moving remaining operations after the cut identity gate to the new qubit; see :ref:`wire cutting as move`.

> There are `three settings <https://www.ibm.com/quantum/blog/circuit-knitting-with-classical-communication>`__ to consider for circuit cutting.  The first is where only local operations (LO) &mdash; as in, local *quantum* operations &mdash; are available.  The other settings introduce classical communication between the circuit executions, which is known in the quantum information literature as LOCC, for `local operations and classical communication <https://en.wikipedia.org/wiki/LOCC>`__.  The LOCC can be either near-time, one-directional communication between the circuit executions (the second setting), or real-time, bi-directional communication (the third setting).

Reviewer: will render by Sphinx as an mdash instead of literally as the text `&mdash;`

```suggestion
There are `three settings <https://www.ibm.com/quantum/blog/circuit-knitting-with-classical-communication>`__ to consider for circuit cutting.  The first is where only local operations (LO) --- as in, local *quantum* operations --- are available.  The other settings introduce classical communication between the circuit executions, which is known in the quantum information literature as LOCC, for `local operations and classical communication <https://en.wikipedia.org/wiki/LOCC>`__.  The LOCC can be either near-time, one-directional communication between the circuit executions (the second setting), or real-time, bi-directional communication (the third setting).
```

A wire cut is represented fundamentally by this package as a two-qubit :class:`.Move` instruction, which is defined as a reset of the second qubit followed by a swap of both qubits.  Equivalently, the operation is defined as transferring the state of the first qubit wire to the second qubit wire, while simultaneously discarding the state of the second qubit wire (the first qubit ends up in state :math:`\lvert 0 \rangle`).

> We have chosen to represent wire cuts in this way primarily because it is consistent with the required way to treat wire cuts when acting on physical qubits: for instance, a wire cut might take the state of physical qubit :math:`n` and continue it as physical qubit :math:`m` after the cut.  Our choice also has the benefit of allowing us to think of "instruction cutting" as a unified framework for considering both wire cuts and gate cuts in the same formalism, because a wire cut is just a cut :class:`.Move` instruction.

Reviewer: ```suggestion
We have chosen to represent wire cuts in this way primarily because it is consistent with the mental model for treating wire cuts when acting on physical qubits: for instance, a wire cut might take the state of physical qubit :math:`n` and continue it as physical qubit :math:`m` after the cut.  Our choice also has the benefit of allowing us to think of "instruction cutting" as a unified framework for considering both wire cuts and gate cuts in the same formalism, because a wire cut is just a cut :class:`.Move` instruction.
```

More information on this formalism is given in Sec. 3 of Ref. \[`4 <https://arxiv.org/abs/2302.03366>`__\].

If you prefer to place cut wires abstractly on a single-qubit wire, please see the `how-to guide on placing wire cuts using a single-qubit instruction <../how-tos/how_to_specify_cut_wires.ipynb>`__, which explains how to use the :func:`.cut_wires` function to convert a circuit with :class:`.CutWire` instructions to a circuit with :class:`.Move`\ s on additional qubits.

Sample weights in the Qiskit addon for circuit cutting
------------------------------------------------------

> In this package, the number of samples taken from the distribution is generally controlled by a ``num_samples`` argument, and each sample has an associated weight that is used during expectation value reconstruction. Each weight with absolute value above a threshold of 1 / ``num_samples`` will be evaluated exactly.  The remaining low-probability elements &mdash; those in the tail of the distribution &mdash; will then be sampled, resulting in at most ``num_samples`` unique weights. Setting ``num_samples`` to infinity indicates that all weights should be generated rigorously, rather than by sampling from the distribution.

Reviewer: ```suggestion
In this package, the number of samples taken from the distribution is generally controlled by a ``num_samples`` argument, and each sample has an associated weight that is used during expectation value reconstruction. Each weight with absolute value above a threshold of 1 / ``num_samples`` will be evaluated exactly.  The remaining low-probability elements --- those in the tail of the distribution --- will then be sampled, resulting in at most ``num_samples`` unique weights. Setting ``num_samples`` to infinity indicates that all weights should be generated rigorously, rather than by sampling from the distribution.
```

> Much of the circuit cutting literature describes a process where we sample from the distribution and take a single shot, then sample from the distribution again and repeat; however, this is not feasible in practice, so we instead perform all sampling upfront.  For now, because of limitations in version 1 of the Qiskit primitives, we take a fixed number of shots for each considered subexperiment and send the subexperiments to the backend(s) in batches. During reconstruction, each subexperiment contributes to the final result with proportion equal to its weight.  We must ensure the number of shots taken is sufficient for the heaviest weighted subexperiment.  In the future, we plan to support passing an individual ``shots`` count with each subexperiment to Qiskit Runtime, so that each subexperiment will be run with a number of shots proportional to that subexperiment's weight in the final result (see issue `#532 <https://github.com/Qiskit/qiskit-addon-cutting/issues/532>`__).  This per-experiment shots count is a new feature enabled by version 2 of the Qiskit primitives.

Author: Does this require any updating in:re primitives, or is this all still true?

Reviewer: The primitives now support this, but this addon still isn't able to leverage that feature -- that's what the issue is about.  So I think the wording should be updated to reflect that the primitives v2 have the necessary feature, and there is an open issue to allow the addon to leverage this.  I suggest removing the phrase "In the future, we plan to support" because it is not clear that anyone has plans at this point to fix this issue.

By the way, I expect that the `&mdash;`s will not render properly in Sphinx/restructedtext.  Would you be open to using either three contiguous hyphens (`---`) or a unicode mdash?

Author: I think unicode will work but I'm not sure how translation will deal with the three hyphens (which is why I have been using `&mdash;` in the first place). Will we still be running sphinx once the docs have migrated?

Reviewer: @eharvey328 can clarify -- but my understanding is that the documentation pipeline will run Sphinx and use its output as input to a later stage of the build.  Assuming this is correct, then `---` will be converted to the mdash unicode character when Sphinx runs.

Author: Ah I see! That would indeed make things easier. @eharvey328 would `--` also get converted to emdash in unicode?

Reviewer: I believe that Sphinx will convert `--` to an ndash instead.

> Much of the circuit cutting literature describes a process where we sample from the distribution and take a single shot, then sample from the distribution again and repeat; however, this is not feasible in practice, so we instead perform all sampling upfront.  For now, because of limitations in version 1 of the Qiskit primitives, we take a fixed number of shots for each considered subexperiment and send the subexperiments to the backend(s) in batches. During reconstruction, each subexperiment contributes to the final result with proportion equal to its weight.  We must ensure the number of shots taken is sufficient for the heaviest weighted subexperiment.  In the future, we plan to support passing an individual ``shots`` count with each subexperiment to Qiskit Runtime, so that each subexperiment will be run with a number of shots proportional to that subexperiment's weight in the final result (see issue `#532 <https://github.com/Qiskit/qiskit-addon-cutting/issues/532>`__).  This per-experiment shots count is a new feature enabled by version 2 of the Qiskit primitives.

Author: ```suggestion
Much of the circuit cutting literature describes a process where we sample from the distribution and take a single shot, then sample from the distribution again and repeat; however, this is not feasible in practice, so we instead perform all sampling upfront.  For now, because of limitations in version 1 of the Qiskit primitives, we take a fixed number of shots for each considered subexperiment and send the subexperiments to the backend(s) in batches. During reconstruction, each subexperiment contributes to the final result with proportion equal to its weight.  We must ensure the number of shots taken is sufficient for the heaviest weighted subexperiment.  An `open issue (#532) <https://github.com/Qiskit/qiskit-addon-cutting/issues/532>`__ tracks work to support passing an individual ``shots`` count with each subexperiment to Qiskit Runtime, so that each subexperiment will be run with a number of shots proportional to that subexperiment's weight in the final result.  This per-experiment shots count is a new feature enabled by version 2 of the Qiskit primitives.
```

Author: @garrison how is this rewording?

Reviewer: It's great! Thank you
